src/communication/response_broker.hpp

```diff
@@ -166,11 +185,28 @@ namespace cubcomm
   template <typename T_PAYLOAD, typename T_ERROR>
   response_broker<T_PAYLOAD, T_ERROR>::bucket::~bucket ()
   {
-    // NOTE: might not hold in the event that a peer server crashes before consuming a response;
-    // in which case the response will linger on in the queue indefinitely
+    // NOTE:
+    //  - might not hold in the event that a peer server crashes before providing a response;
+    //  - currently this is either implemented (or not) in the application layers above;
+    //  - if it were to be sorted out at this level, a timeout must be added after which an
+    //    error is returned to the waiting thread
     assert (m_response_payloads.empty ());
   }
 
+  template <typename T_PAYLOAD, typename T_ERROR>
+  void
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::register_request (response_sequence_number a_rsn)
+  {
+    {
+      std::lock_guard<std::mutex> lk_guard (m_mutex);
```

Comment:
What happens if `register_request()` is called while `notify_terminate_and_wait()` is waiting with the lock of m_mutex until the last request is digested? It seems that this will acquire the lock and register the request after `notify_terminate_and_wait()` confirms the container is empty and exits. Then, `m_response_payloads` ends up having some requests left.
```suggestion
      std::lock_guard<std::mutex> lk_guard (m_mutex);
      if (m_terminate)
      {
        return;
      }
```

Reply:
Bad things can happen.
However, I do not think it is the responsibility of this lower application layer to deal with this.
It would be the job of the connection handlers (instantiated by `page_server` and `tran_server` objects) to deal with the order of tearing down a connection.
Somewhere we have to draw a line. Not all layers have to deal with all aspects.
For consistency though, we can at least add an assert.
What do you think?

Reply:
Fair enough. Let's go with just an assert. 
> Somewhere we have to draw a line. Not all layers have to deal with all aspects. 

Good pointer. Thank you.